sphinx/builders/linkcheck.py

```diff
@@ -706,15 +713,37 @@ def contains_anchor(response: Response, anchor: str) -> bool:
 class AnchorCheckParser(HTMLParser):
     """Specialised HTML parser that looks for a specific anchor."""
 
-    def __init__(self, search_anchor: str) -> None:
+    def __init__(self, search_anchor: str, original_encoded_anchor: str = '') -> None:
+        """Initialize the parser with multiple anchor variations.
+
+        Args:
+            search_anchor: The decoded anchor to search for
+                (e.g., "standard-input/output-stdio")
+            original_encoded_anchor: The original encoded anchor
+                (e.g., "standard-input%2Foutput-stdio")
+        """
         super().__init__()
 
-        self.search_anchor = search_anchor
+        # Create variations of the anchor to check
+        self.search_variations = {
+            search_anchor,  # decoded (current behavior)
+        }
+
+        # Add the original encoded version if provided
+        if original_encoded_anchor:
+            self.search_variations.add(original_encoded_anchor)
+
+        # Add a re-encoded version if the decoded anchor contains characters
+        # that would be encoded
+        if search_anchor != quote(search_anchor, safe=''):
+            self.search_variations.add(quote(search_anchor, safe=''))
```

Comment:
It would be nice to extract the anchor-to-search-variations code into a helper function/method; doing so would hopefully also be convenient to write unit tests around its behaviour.